MRG: Whitening in report

mainakjas opened the pull request:
closes https://github.com/mne-tools/mne-python/issues/1730

almost ready. Just needs a few final touches.

Eric89GXL commented on mne/report.py:
```diff
@@ -1130,6 +1156,22 @@ def _render_toc(self, verbose=None):
                         html_toc += u'</ul></li>'
 
+                    if fname.endswith('(whitened)'):
+                        # XXX: remove redundant read_evokeds
```
This shouldn't really matter too much, reading evokeds is fast. If it makes the code uglier to re-use, I wouldn't bother.

Eric89GXL commented on mne/report.py:
```diff
@@ -645,6 +665,8 @@ class Report(object):
     info_fname : str
         Name of the file containing the info dictionary.
+    cov_fname : str
+        Name of the file containing the noise covariance.
```
This has to go at the end (before `verbose`) to avoid breaking people's code. `Report` has been in mne-python for a while now...

dengemann commented on mne/report.py:
```diff
@@ -1414,6 +1456,33 @@ def _render_cov(self, cov_fname, info_fname):
+            noise_cov = regularize(noise_cov, ev.info)
+            ev_white = whiten_evoked(ev, noise_cov, diag=True)
+            global_id = self._get_id()
+
```
This is not the right way to do it. 
Use `evoked.plot_white(cov)` instead.

mainakjas replied:
`show` doesn't work for that function and @teonlamont also has the issue and we can't figure out why :)

Eric89GXL replied:
You mean if you load an evoked and a cov, and do `evoked.plot_white(cov, show=True)` nothing pops up?

mainakjas replied:
no, the other way round. If I do `show=False`, something pops up. We don't want that in the report.

mainakjas replied:
never mind, I found a fix

dengemann replied:
@mainakjas then let's fix it and figure it out, all the logic that's baked into evoked.plot_white really provides a more adequate workflow.

dengemann replied:
if you have SSS'd data what you do here is even more or less wrong.

mainakjas replied:
I still have some weird issue. Try this:

```py
from mne import read_evokeds, Covariance
from mne.datasets import sample

path = sample.data_path()
cov_fname = path + '/MEG/sample/sample_audvis-cov.fif'

noise_cov = Covariance(cov_fname)

evoked = read_evokeds(path + '/MEG/sample/sample_audvis-no-filter-ave.fif')
evoked[0].plot_white(noise_cov)
```

dengemann replied:
'method' is not defined in the dict if the cov is not freshly computed.


option 1) if 'method' not in cov method = 'unknown'
option 2) save the 'method' tag

Btw. we have a baselining issue with that particular example.

Eric89GXL replied:
I think if the method is unknown, it can safely assume that it is `empirical`, at least for the purposes of plotting. And that probably needs `read_evokeds(..., baseline=(None, 0))` to fix the baseline.

dengemann replied:
Would always be the wrong assumption on my case, as I'm using the 'auto' option and saving stuff to disk.

dengemann replied:
So on the long run we might want to store that info.

Eric89GXL replied:
Oh, we don't save the method to disk?

dengemann replied:
> And that probably needs read_evokeds(..., baseline=(None, 0)) to fix the baseline.

we had that discussion before. By default the report will not apply any baselining, therefore the results look are really difficult to parse very often.

dengemann replied:
> Oh, we don't save the method to disk?

apparently not.

dengemann replied:
ahh yes we do.

dengemann replied:
we should change this and add 'empirical' if no method is written
https://github.com/mne-tools/mne-python/blob/master/mne/cov.py#L1533

Eric89GXL replied:
Maybe we should add a parameter saying whether or not baselining should be done, otherwise the whitening plot will look like crap. Or at least have it on by default for the plot_white part (and not the others).

dengemann replied:
The problem is that you don't really know of you need it. Maybe as a rule of thumb if highpass < 1

Eric89GXL replied:
Sure, we could have a parameter `True/False/'auto'`, if `auto` use the highpass information.

-1 on changing the reading default, that will make it so that if I do `write_cov(read_cov())` I get something added, which isn't great. I think we should just change our logic functions. It's more work but a little bit cleaner from an I/O perspective.

dengemann replied:
As we have the writing we can just assume that if 'method' not in noise_cov it must be 'empirical'.

Eric89GXL replied:
Exactly, if we have actually done something other than empirical, it will not be None and it not be "empirical", so our check can be `if cov['method'] is None or cov['method'] == 'empirical'`

dengemann replied:
but the field is inexsistant if it has not been stored.

dengemann replied:
```Python

if 'method' not in cov:
   method = 'empirical'
```

Therefore

Eric89GXL replied:
`method = cov.get('method', 'empirical')`

dengemann replied:
yeah, exactly.

mainakjas replied:
perfect. Thanks for the discussion. I learnt something :) This fixes it

mainakjas replied:
more surprises @dengemann : 

```py
from mne import read_evokeds, Covariance
from mne.datasets import sample

path = sample.data_path()
cov_fname = path + '/MEG/sample/sample_audvis-cov.fif'

noise_cov = Covariance(cov_fname)

evoked = read_evokeds(path + '/MEG/sample/sample_audvis-ave.fif')
evoked[0].plot_white(noise_cov)
```

gives

![gfp_issue](https://cloud.githubusercontent.com/assets/3817535/7264722/3e9fd354-e88f-11e4-8d03-b4af3d1b47c7.png)

if you know anything off the top of your head let me know. Otherwise, I'll try to figure it out tomorrow.

dengemann replied:
Well, this is a well calibrated diagnostic plot, it shows you that
something isn't right :) That's expected crap, the data is not high pass
filtered at 1 Hz. We need to consider the discussion we had with Eric. If
not filtered, baseline data.

dengemann commented on mne/report.py:
```diff
@@ -136,6 +137,9 @@ def _is_bad_fname(fname):
        and highlighting them in red in the TOC.
     """
+    if fname.endswith('(whitened)'):
+        fname = fname[:-11]
```
why is this actually called fname, it's the HTML variables, right, no files involved?

mainakjas replied:
just historical reasons. Initially, we didn't have custom plots -- only files in the table of contents. We could rename it at some point.

agramfort replied:
+1 for rename if it helps readability

dengemann commented on mne/report.py:
```diff
@@ -1414,6 +1458,32 @@ def _render_cov(self, cov_fname, info_fname):
+
+        html = []
+        for ev in evokeds:
+            noise_cov = regularize(noise_cov, ev.info)
```
This is not a good idea. please remove the regularization. we should show things how they are. Our cov estimation now takes care of this.

dengemann commented on mne/report.py:
```diff
@@ -994,12 +1017,17 @@ def parse_folder(self, data_path, pattern='*.fif', n_jobs=1, mri_decim=2,
             info, sfreq = None, None
 
+        if self.cov_fname is not None:
+            cov = Covariance(self.cov_fname)
```
why not the reader mne.read_cov?

Eric89GXL commented:
@dengemann and I talked about doing baseline correction by default if `highpass is None or highpass < 1.`, that might fix the weird levels

Eric89GXL commented:
(baseline correcting only for the plot_white stuff, not for normal plotting)

dengemann commented:
@mainakjas @Eric89GXL this his how the world looks like after baselining (on master):

```Python
from mne import read_evokeds, Covariance
from mne.datasets import sample

path = sample.data_path()
cov_fname = path + '/MEG/sample/sample_audvis-cov.fif'

noise_cov = Covariance(cov_fname)
noise_cov['method'] = 'empirical'
evoked = read_evokeds(path + '/MEG/sample/sample_audvis-ave.fif', 
                      baseline=(None, 0))

evoked[0].plot_white(noise_cov)
```
![screenshot 2015-04-22 01 38 42](https://cloud.githubusercontent.com/assets/1908618/7264798/6c8d36f2-e890-11e4-8715-bd5ec44b0d1d.png)

dengemann commented:
Btw I think this also nicely shows spurious late components produced by baselining (instead of filtering).

mainakjas commented:
so do we just go with `auto` option? I am afraid adding another param to the command line may make things clunky.

dengemann commented:
> so do we just go with auto option? I am afraid adding another param to the command line may make things clunky.

well but better clunky than wrong. Maybe we can make it a default choice.

mainakjas commented:
cool, this working good with 'auto' baseline. I'll refactor some of the code tomorrow so that we have fewer green lines and then it'll be good to go!

agramfort commented on mne/commands/mne_report.py:
```diff
@@ -61,9 +68,17 @@ def run():
+    elif options.baseline == 'False':
+        baseline = False
+    elif options.baseline == 'auto':
+        baseline = 'auto'
```
is this an API change?

agramfort commented on mne/report.py:
```diff
@@ -652,14 +674,22 @@ class Report(object):
+        Name of the file containing the noise covariance.
+    baseline : str | bool
+        If True, apply baseline. If 'auto', baseline correct only if highpass
+        filter has not been applied or is less than 1.0 Hz.
```
It's maybe black magic and not consensual....

dengemann replied:
@agramfort any better idea how to avoid the baseline mess by default? At least this is magic for a good purpose.

dengemann replied:
howver we should do more than just True | str. See regular baseline API.

agramfort replied:
I'd like people to save baseline corrected data. So we avoid this magic. To
me it's a mistake that MNE delays baselining at inverse modeling /
mne_analyze stage.

I'd go for --bmin and --bmax options to set baseline and avoid magic.

Eric89GXL replied:
I guess in order not to break backward compat it should be `baseline=None` by default here, and add `bmin` and `bmax` to the command-line interface.

agramfort replied:
+1

On Wed, Apr 22, 2015 at 7:01 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/report.py
> <https://github.com/mne-tools/mne-python/pull/1993#discussion_r28892884>:
>
> > @@ -652,14 +674,22 @@ class Report(object):
> >          Subject name.
> >      title : str
> >          Title of the report.
> > +    cov_fname : str
> > +        Name of the file containing the noise covariance.
> > +    baseline : str | bool
> > +        If True, apply baseline. If 'auto', baseline correct only if highpass
> > +        filter has not been applied or is less than 1.0 Hz.
>
> I guess in order not to break backward compat it should be baseline=None
> by default here, and add bmin and bmax to the command-line interface.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1993/files#r28892884>.
>

teonlamont commented on mne/viz/evoked.py:
```diff
@@ -426,7 +426,7 @@ def _plot_evoked_white(evoked, noise_cov, scalings=None, rank=None, show=True):
     axes_evoked = None
 
-    def whitened_gfp(x, rank=None):
+    def whitened_gfp(x, rank=None, show=True):
```
show isn't needed here

mainakjas replied:
are your fixes merged now?

teonlamont replied:
almost, i think Travis just stalled

teonlamont replied:
btw, it's merged

teonlamont commented on mne/viz/evoked.py:
```diff
@@ -492,14 +492,14 @@ def whitened_gfp(x, rank=None):
 
             data = evoked_white.data[sub_picks]
-            gfp = whitened_gfp(data, rank=this_rank)
+            gfp = whitened_gfp(data, rank=this_rank, show=show)
```
nor here

agramfort commented:
you need to rebase. Anything else to do here?

dengemann commented:
Some comments by Teon are not addressed yet.

2015-04-26 10:19 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

> you need to rebase. Anything else to do here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1993#issuecomment-96342441>.
>

mainakjas commented:
I'm on it.

mainakjas commented:
@teonlamont I addressed your comments, but something is odd after rebasing to master. The figures in the `gfp` plot start showing up again. Are you sure it was fixed?

Eric89GXL commented:
@mainakjas good to go from your end?

Eric89GXL commented on doc/source/mne_report_tutorial.rst:
```diff
@@ -45,6 +45,12 @@ To properly render `trans` and `covariance` files, add the measurement informati
+    
+    mne report --path MNE-sample-data/ --info MNE-sample-data/MEG/sample/sample_audvis-ave.fif \ 
+        --cov MNE-sample-data/MEG/sample/sample_audvis-cov.fif --subject sample \
+        --subjects-dir MNE-sample-data/subjects --verbose
```
Won't this need `--bmin` and `--bmax` to render properly? I don't think `sample_audvis-ave.fif` is saved baselined. Did you try this line?

Eric89GXL replied:
Yeah I think it needs it:

![screenshot from 2015-04-27 10 40 13](https://cloud.githubusercontent.com/assets/2365790/7353624/d8e8ab5a-ecc9-11e4-8072-6a8e1a05deee.png)

mainakjas replied:
okay, let me fix this

Eric89GXL commented:
FYI I killed your CI builds to make room for others since I assume you need to make the change above.

mainakjas commented:
docs updated @Eric89GXL

Eric89GXL commented:
Much better:

![screenshot from 2015-04-27 11 48 39](https://cloud.githubusercontent.com/assets/2365790/7354945/6cb1b882-ecd3-11e4-97cb-d220bf58fb57.png)

I'll merge once Travis is happy unless anyone else speaks up in the meantime.

Eric89GXL commented:
Actually, we probably want the baseline to apply to all plots, not just the whitened one:

![screenshot from 2015-04-27 11 50 24](https://cloud.githubusercontent.com/assets/2365790/7354967/a1bd24c6-ecd3-11e4-8f21-3b7a19a97160.png)

WDYT?

mainakjas commented:
On Mon, Apr 27, 2015 at 8:50 PM, Eric Larson <notifications@github.com>
wrote:

> Actually, we probably want the baseline to apply to all plots, not just
> the whitened one:
>
> [image: screenshot from 2015-04-27 11 50 24]
> <https://cloud.githubusercontent.com/assets/2365790/7354967/a1bd24c6-ecd3-11e4-8f21-3b7a19a97160.png>
>
> WDYT?
>
I agree but I think there was some discussion btw @agramfort and @dengemann
about this and the consensus was that only whitened plots should get the
baseline, no?

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1993#issuecomment-96777999>.
>

Eric89GXL commented:
I think that was back when we were talking about auto-baselining, I think if a user passes a baseline it should go for every evoked. @dengemann @agramfort do you agree?

agramfort commented:
Yes baseline everything 


> On 27 avr. 2015, at 20:56, Eric Larson <notifications@github.com> wrote:
> 
> I think that was back when we were talking about auto-baselining, I think if a user passes a baseline it should go for every evoked. @dengemann @agramfort do you agree?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mainakjas commented:
alright, we baseline everything now. It's good to go from my end @Eric89GXL

Eric89GXL commented on mne/viz/topomap.py:
```diff
@@ -1085,7 +1085,8 @@ def plot_evoked_topomap(evoked, times=None, ch_type='mag', layout=None,
                               outlines=outlines, image_mask=image_mask,
-                              contours=contours, image_interp=image_interp)
+                              contours=contours, image_interp=image_interp,
+                              show=show)
```
`show=False` so the showing gets delayed until the end of the function call on line 1125

mainakjas replied:
ah, now I get it. Otherwise, the figure will pop up multiple times when `show=True`

Eric89GXL commented:
One last tweak @mainakjas, then I'll merge once the CIs are happy.

mainakjas commented:
taken care of!

Eric89GXL commented:
Thanks @mainakjas!

dengemann commented:
Thanks @mainakjas !!! A beer-stew on me.

2015-04-28 1:18 GMT+02:00 Eric Larson <notifications@github.com>:

> Thanks @mainakjas <https://github.com/mainakjas>!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1993#issuecomment-96848452>.
>

agramfort commented on doc/source/mne_report_tutorial.rst:
```diff
@@ -45,11 +45,21 @@ To properly render `trans` and `covariance` files, add the measurement informati
         --subject sample --subjects-dir MNE-sample-data/subjects --verbose
 
+To render whitened `evoked` files with baseline correction, add the noise covariance file::
+    
```
You mention baseline but I don't see it

mainakjas replied:
It's `--bmax 0`. It's there :)

agramfort replied:
My bad

agramfort commented:
Please fix in master. Thanks heaps @mainakjas !
